sgo/types/lookup.go

```diff
@@ -294,6 +293,36 @@ func assertableTo(V *Interface, T Type) (method *Func, wrongType bool, mustOptio
 	return
 }
 
+// typeMustOptional reports whether a type T recursively needs to be wrapped in an optional.
+// If T is Array or Slice, the underlying type will be checked. If T is Optional, it will not
+// need to be wrapped in an optional unless the underlying type is a Map. In that case, further
+// checks are needed for the key and value type of the map.
+func typeMustOptional(T Type) bool {
+	t := T.Underlying()
+	switch t.(type) {
+	case *Pointer, *Map, *Signature, *Chan, *Interface:
+		return true
+	case *Slice, *Array:
+		var underlying Type
+		switch t.(type) {
+		case *Slice:
+			underlying = t.(*Slice).Elem()
+		case *Array:
+			underlying = t.(*Array).Elem()
+		}
+
+		return typeMustOptional(underlying)
+	case *Optional:
+		switch t.(*Optional).Elem().(type) {
+		case *Map:
```

Comment:
If something is wrapped in an optional, the only ones that can have another type to possibly require an optional are map, slice and array. Array and Slice can't be wrapped in optionals. So the only other check that needs to be done is in map, that's why the switch only covers that case. The rest don't have types that need to be checked and they are themselves wrapped in an optional so they don't need to be wrapped for sure.